packages/editor/src/services/container.service.ts

```diff
@@ -0,0 +1,266 @@
+            }
+          };
+        } else {
+          schema = {
+            ...schema, ...{
+              [propertyRoot]: { [property]: parentSchema[propertyRoot][property] }
+            }
+          };
+        }
+      }
+    });
+  }
+
+  return schema;
+};
+
+/**
+ * Retieves a self contained schema.
+ * @param parentSchema The schema to use for resolvement
+ * @param refPath The path to resolve
+ * @return a JsonSchema that is self-contained
+ */
+export const getSelfContainedSchema = (parentSchema: JsonSchema, refPath: string): JsonSchema => {
+  let schema = resolveSchema(parentSchema, refPath);
+  schema = deepCopy(schema);
+  if (_.isEmpty(schema.id)) {
+    schema.id = '#' + refPath;
+  }
+  schema = { ...schema, ...calculateSchemaReferences(parentSchema, schema) };
+
+  return schema;
+};
+
+/**
+ * Calculates label for an item which is resolved in anyOf
+ * @param {string} id The schema id
+ * @returns {string}
+ */
+const calculateLabel = (id: string): string => (
+  _.startsWith(id, '#') ? id.substring(id.lastIndexOf('#') + 1) : id
+);
+
+/**
+ * Constructing container properties of a given schema
+ * @param {string} property The schema key from which this property was created.
+ * @param {string} label The name of the schema the property describes.
+ * @param {JsonSchema} schema The schema is the JsonSchema this property describes.
+ * @param {JsonSchema} rootSchema The parent schema
+ * @param {boolean} isInContainment To indicate whether the schema is in a container or not
+ * @param {boolean} internal To accept only directly contained containers
+ * @returns {Property[]}
+ */
+export const getContainerProperty = (property: string,
+                                     label: string,
+                                     schema: JsonSchema,
+                                     rootSchema: JsonSchema,
+                                     isInContainment: boolean,
+                                     internal = false): Property[] => {
+  if (schema.$ref !== undefined) {
+    return getContainerProperty(
+      property,
+      schema.$ref === '#' ? undefined : schema.$ref.substring(schema.$ref.lastIndexOf('/') + 1),
+      getSelfContainedSchema(rootSchema, schema.$ref),
+      rootSchema,
+      isInContainment,
+      internal
+    );
+  }
+  if (isObject(schema)) {
+    if (isInContainment) {
+      const prop: Property =  {
+        property,
+        label,
+        schema
+      };
+      return [prop];
+    }
+    if (internal) {
+      // If internal is true the schema service does not need to resolve further,
+      // because only directly contained containments are wanted.
+      // This prevents running into circles
+      return [];
+    }
+
+    return Object.keys(schema.properties)
+      .reduce(
+        (prev, cur) =>
+          prev.concat(
+            getContainerProperty(
+              cur,
+              cur,
+              schema.properties[cur],
+              rootSchema,
+              false,
```

Comment:
Is passing `false` as the value for `isInContainment` not contradicting its own name here? Also why is `internal` set to `true`? These seem like arbitrary choices. Can you explain? Maybe update the JsDoc accordingly while you are at it

Reply:
I will update JsDoc. `isInContainment` is set to true if the schema has items and these items are non primitive properties, so the term `container` is used for these property objects. So this container property is directly related to type `array`. The term `property` is quite confusing but I don't know which term is correct to define this logic. The logic itself makes sense since we add children into parent so we actually need to consider parents as `containers`. `internal` value also depends on this logic. If the schema type is object, then we don't actually need to resolve it further because we already know the results from future iterations will be the children of the corresponding schema

Reply:
Maybe we should use the term `getChildren` instead of `getProperties`